File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(vector<int>& A, int& i, int bound) {
        if(i >= A.size() || A[i] > bound) return NULL;
        TreeNode* root = new TreeNode(A[i++]);
        root->left = dfs(A, i, root->val);
        root->right = dfs(A, i, bound);
        return root;
    }
    TreeNode* bstFromPreorder(vector<int>& A) {
        int i = 0;
        return dfs(A, i, INT_MAX);
    }
};

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, A: List[int]) -> Optional[TreeNode]:
        def helper(i, j):
            if i == j: return None
            node = TreeNode(A[i])
            mid = bisect.bisect(A, A[i], i+1, j)
            node.left = helper(i+1, mid)
            node.right = helper(mid, j)
            return node
        return helper(0, len(A))